### PR TITLE
Add Delete Volume Finalizer

### PR DESCRIPTION
**Pull request by xing-yang**
This PR adds Finalizer on the snapshot source PVC to prevent it
from being deleted when a snapshot is being created from it.

### CODE REVIEW

**Comment by xing-yang** [comment]
@wackxu PTAL.  Thanks.

**Comment by wackxu** [comment]
I think we should not  depend on annotation `SnapshotsInCreationCountAnnotation` on pvc to judge  the pvc is being in used by snapshot.
1、we should also consider snapshot that user created with non-empty `SnapshotContentName` .
2、when a snapshot is deleted, we should also decrease the `SnapshotsInCreationCountAnnotation`
3、if `SnapshotsInCreationCountAnnotation` is increased when create snapshot but then the controller is  failed and another controller get the lead election, then the  `SnapshotsInCreationCountAnnotation`  will increased again.

**Comment by wackxu** [comment]
We can traverse the snapshot in the cache to determine if pvc is being used. like the pvc_protection_controller in the k8s.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L210-L237

**Comment by xing-yang** [comment]
> We can traverse the snapshot in the cache to determine if pvc is being used. like the pvc_protection_controller in the k8s.
> https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L210-L237

This seems to be a good idea.  I'll explore it.  Thanks.

**Comment by xing-yang** [comment]
Rebased.

**Comment by xing-yang** [comment]
@wackxu Addressed your comments.  Use snapshotLister to check if a snapshot is being created from a PVC.

**Comment by xing-yang** [comment]
@jsafrane PTAL.  Thanks.

File: pkg/controller/snapshot_controller.go [hunk 1]
```diff
@@ -1068,3 +1089,117 @@ func (ctrl *csiSnapshotController) removeSnapshotFinalizer(snapshot *crdv1.Volum
 	klog.V(5).Infof("Removed protection finalizer from volume snapshot %s", snapshotKey(snapshot))
 	return nil
 }
+
+// addSnapshotSourceFinalizer adds a Finalizer for VolumeSnapshot Source PVC.
+func (ctrl *csiSnapshotController) addSnapshotSourceFinalizer(snapshot *crdv1.VolumeSnapshot) error {
```
Comment (jsafrane) [inline pkg/controller/snapshot_controller.go]: I think you can merge `addSnapshotSourceFinalizer` and `needToAddSnapshotSourceFinalizer` together into e.g. `ensureSnapshotSourceFinalizer` and save one `getClaimFromVolumeSnapshot` and couple of common lines of code.

Comment (xing-yang) [reply pkg/controller/snapshot_controller.go]: Sure.  Merged the two functions into one.

**Comment by xing-yang** [comment]
Addressed one comment.  Will add tests soon.

**Comment by xing-yang** [comment]
Hi @jsafrane, I added tests. PTAL.

**Comment by xing-yang** [comment]
e2e test failed because external-snapshotter does not have "update" rbac rule for updating PVC object to add a Finalizer:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/testing-manifests/storage-csi/external-snapshotter/rbac.yaml

Error message in e2e tests:
```
Apr 25 02:24:54.933: INFO: At 2019-04-25 02:14:52 +0000 UTC - event for snapshot-xxdgm: {csi-snapshotter csi-hostpath } SnapshotCreationFailed: Failed to create snapshot: snapshot controller failed to update pvc-flxrc on API server: persistentvolumeclaims "pvc-flxrc" is forbidden: User "system:serviceaccount:default:csi-snapshotter" cannot update resource "persistentvolumeclaims" in API group "" in the namespace "provisioning-7646"
```

**Pull request merged**